tests/tests.hs

```diff
@@ -1,44 +1,145 @@
+      withSystemTempDirectory "rocksdb" $ \path -> do
+        withDB path testOpenOptions $ \db -> do
+          put db def "zzz" "zzz"
+          val <- get db def "zzz"
+          return val
+      `shouldReturn` (Just "zzz")
+
+    it "should put items into the database and retrieve them (non-async-safe way)" $ do
+      withSystemTempDirectory "rocksdb" $ \path -> do
+        db <- open path testOpenOptions
         put db def "zzz" "zzz"
-        get db def "zzz"
+        val <- get db def "zzz"
+        close db
+        return val
       `shouldReturn` (Just "zzz")
 
-    it "should put items into a database whose filepath has unicode characters and\
-       \ retrieve them" $  do
-      runResourceT $ withSystemTempDirectory "rocksdb" $ \path -> do
+    it "should put items into the database and retrieve them (MonadResource way)" $ do
+      runResourceTChecked $ withSystemTempDirectory "rocksdb" $ \path -> do
+        (_, db) <- openBracket path testOpenOptions
+        put db def "zzz" "zzz"
+        val <- get db def "zzz"
+        return val
+      `shouldReturn` (Just "zzz")
+
+    -- This test doesn't apply to Windows because there we can't delete the
+    -- directory while a file in it is open.
+#if !defined(mingw32_HOST_OS)
+    it "does weird global singleton string matching stuff for double-locking warnings" $ do
+      -- This test will fail when RocksDB fixes this issue
+      -- (see https://stackoverflow.com/questions/37310588/rocksdb-io-error-lock-no-locks-available#comment83145041_37312033).
+      -- It exists so that we get notified when that fixing happens.
+      let path = "/tmp/tmpdir"
+
+      exists <- doesDirectoryExist path
+      when exists $ removeDirectoryRecursive path
+
+      createDirectoryIfMissing True path
+
+      db <- open path testOpenOptions
+      (do
+        put db def "zzz" "zzz"
+        get db def "zzz") `shouldReturn` (Just "zzz")
+      -- We purposely don't `close` the DB above.
+
+      removeDirectoryRecursive path
+
+      open path testOpenOptions `shouldThrow` \ioe ->
+        isUserErrorType (ioeGetErrorType ioe) &&
+        ioeGetErrorString ioe `elem`
+          [ "open: IO error: lock " ++ path ++ "/LOCK: No locks available"
+          , "open: IO error: While lock file: " ++ path ++ "/LOCK: No locks available"
+          ]
+
+      -- Close first DB now so that it doesn't memory-leak.
+      close db
+#endif
+
+    it "should put items into a database whose filepath has unicode characters and retrieve them" $ do
+      runResourceTChecked $ withSystemTempDirectory "rocksdb" $ \path -> do
         unicode <- getUnicodeString <$> liftIO (generate arbitrary)
-        db <- initializeDB $ path ++ unicode
+        (_, db) <- openBracket (path </> "unicode-randomdir-" ++ unicode) testOpenOptions
         put db def "zzz" "zzz"
-        get db def "zzz"
+        val <- get db def "zzz"
+        return val
```

Comment:
Btw, why did you change:
```
get db def "zzz"
```
to
```
        val <- get db def "zzz"
        return val
```

Do you think the latter is more readable? 🤔

Reply:
(It's likely that he was printing or debugging `val` and then didn't come back and return it to the old code.)

Reply:
This actually comes from [here](https://github.com/serokell/rocksdb-haskell/pull/10/commits/fc69de928e8fafb529e571320f88f3f6f4be1978#diff-bfd1c1e5fa80be47287fe256c3cbb01aL59):

I changed to bin the value so that I could insert a `close` in between, and then removed the close again when the `resourcet` patch allowed me to do so:

```diff
         val <- get db def "zzz"
-        close db -- TODO remove; this should make the test fail, but it doesn't because errors in finalizers run by runResourceT don't seem to bubble up
         return val
``` 

We could change it back to just `get db def "zzz"`, I have no opinion on which one is more readable.